rftp/msgs.go

```diff
@@ -99,10 +115,24 @@ func (s *MsgHeader) UnmarshalBinary(data []byte) error {
 	s.msgType = vt & 0x0F
 	s.ackNum = uint8(data[1])
 	s.optionLen = uint8(data[2])
+	if s.optionLen > 0 {
+		s.options = make([]option, s.optionLen)
+	}
 
-	// TODO: Parse options and fix hdrLen
 	s.hdrLen = 3
 
+	lens := data[3:]
+	for i := 0; uint8(i) < s.optionLen; i++ {
+		o := option{}
+		oend := 2 + int(lens[1])
```

Comment:
I think this is kind of mixing up the parsing, because here we already look at the field, which will be parsed in `o.UnmarshalBinary` again. If you just use the open ended slice and trust `o.UnmarshalBinary` to do it's job correctly, you can omit the `oend` variable and keep the parsing separated (see also my comment above about the unreachable error condition).

Reply:
I still need to look at the length of the current option to determine the start of the next option, no? In other words, `oend` is needed to reassign `lens`. It would obviously be better if `UnmarshalBinary` would return the number of consumed byte but that is not the case.

Do you have an idea how this could be solved in a better way?

Reply:
`oend` is equal to `o.length`, after calling `o.UnmarshalBinary`, isn't it?